Add Capability to Inject Secrets, Configuration, Templates and Certificates

estenrye opened the pull request:
Add ability to inject secrets, config & templates.
- Enables users to inject:
  - secrets
  - certificates
  - configuration
  - templates

- Enables users to configure all capabilities of
  step-certificates using only the helm chart.

- Enables users to use the native yaml to configure
  ca.json and have helm automatically translate it
  to json.

Related Issues:
- #45
- #2
- #9
- #22 
- #33

estenrye commented:
Running into issues, going to close this PR and keep poking at it

estenrye commented:
Okay, got the kinks worked out in my load balancer and database configuration.  Should be good for review @maraino.

estenrye commented:
going to work on the documentation for a bit tonight.

estenrye commented:
Added documentation for new values fields.

maraino commented on step-certificates/templates/secrets.yaml:
```diff
@@ -14,3 +14,30 @@ metadata:
+---
+apiVersion: v1
+kind: Secret
+type: smallstep.com/private_keys
```
To be consistent the standard types, we should use `-` instead of `_`.

estenrye replied:
This should be easy to change.

maraino reviewed (changes requested):
Wow, this looks amazing, I've added a small comment, but if you can edit the conditions that display the bootstrap Job, Secrets, and ConfigMaps to depend on the inject variables it will make the users's values.yaml easier.

For example, if the inject of config, certificates and secrets are enabled, you don't need the bootstrap Job. If the inject of config is enabled you don't need the ConfigMap that has the contains the ca.json, ...

I can see that this might get a little bit complicated, we should consider if it just makes sense to have an `inject.enabled` and disables the all the bootstrap resources, and forces you to set all the config, certificates, and secrets. What do you think? 

There might be some users that only want to define a custom ca.json and let the bootstrapper create the secrets. But I think this can cause race conditions where the bootstrapper overwrites the `ca.json`, and unless we add some "ifs" to the script we won't be able to solve those race conditions. So as I said before, we should consider if supporting all the cases makes sense.

estenrye commented:
@maraino thanks for the review and pointing out where I could simplify the experience for the users.

I have updated the pr to use a single enabled flag, rather than three and updated the conditions in the templates such that when `inject.enabled` is true, bootstrap functionality is disabled.

maraino commented on step-certificates/values.yaml:
```diff
@@ -31,6 +36,137 @@ bootstrap:
+
+    ssh:
+      # ssh_host_ca_key contains the contents of your encrypted SSH Host CA key
+      host_ca_key: ""
+      # ssh_host_ca_key: |
```
The comment should be `# host_ca_key: |`

maraino commented on step-certificates/values.yaml:
```diff
@@ -31,6 +36,137 @@ bootstrap:
+      #   -----END OPENSSH PRIVATE KEY-----
+
+      # ssh_user_ca_key contains the contents of your encrypted SSH User CA key
+      user_ca_key: ""
+      # ssh_user_ca_key: |
```
The comment should be `# user_ca_key: |`

maraino reviewed (approved):
I had to make some changes to be able to run the bootstrapper properly.
